### ISSUE TITLE
Determine best simple estimator (no ensemble)

### ISSUE THREAD

**Issue by mattvan83** [EpistasisLab/tpot#553]
Dear experts,

Would it be possible to specify to tpot to determine the best estimator not formed based on ensemble machine learning, i.e. not compound estimators ?

Best regards,
Matthieu

**Comment by weixuanfu** [comment]
Hi Matthieu,

Thank you for interesting question. 

We are thinking about using a configurable grammar (see #523) to specify pipeline structure in future version of TPOT.  With the grammar, we can easily specify simple estimator or pipeline (compound estimators).

**Comment by rhiever** [comment]
Hi @mattvan83,

You can use a [custom TPOT configuration](rhiever.github.io/tpot/using/#customizing-tpots-operators-and-parameters) to constrain the ML algorithms that TPOT optimizes over. The [TPOT light](http://rhiever.github.io/tpot/using/#built-in-tpot-configurations) configuration is a built-in configuration that avoids complex ensemble models, so that is another option.

**Comment by mattvan83** [comment]
Hi,

Thank you for this answer. However, using a custom TPOT configuration needs to specify all ML algorithms wanted to be tested. Wouldn’t it be an option that tests all « simple »  ML algorithms existing ?

TPOT light avoids complex ensemble models, but I would like to simply avoid ensemble models.

**Comment by weixuanfu** [comment]
@rhiever Custom configuration is indeed a good option for specifying one algorithm in one TPOT fit(). But the builtin `StackingEstimator` sometime can build some complex pipelines with even one algorithm.

**Comment by weixuanfu** [comment]
@mattvan83 do you mean that  "ensemble models" is the pipeline?

**Comment by mattvan83** [comment]
@Weixuan What StackingEstimator correspond to if ensemble ML algorithms have been deactivated ?

**Comment by mattvan83** [comment]
@rhiever No I don’t want to suppress preprocessing or normalization but only combination of ML algorithms.

**Comment by weixuanfu** [comment]
StackingEstimator is meta-transformer for adding predictions and/or class probabilities as synthetic feature(s) (see #278)

**Comment by rhiever** [comment]
@mattvan83, do you mean you want to avoid stacking of ML algorithms? For example, a pipeline that performs StandardScaler --> LogisticRegression --> SVM?

Or do you simply mean you want to avoid algorithms that are ensemble models, such as Random Forest and Gradient Tree Boosting?

**Comment by mattvan83** [comment]
@rhiever yes I want to avoid stacking of multiple ML algorithms.

**Comment by weixuanfu** [comment]
I think we still need grammar for this case, like the number of classifier/regressor = 1 in the configurable grammar. I will look into it.

**Comment by rhiever** [comment]
Oh, I see. It's not possible to avoid that in the current version of TPOT, @mattvan83. However, you can access the `pareto_front_fitted_pipelines_` attribute of the TPOT object to see the entire Pareto front of pipelines for the problem. In our case, the Pareto front shows the best-discovered pipelines that have a trade-off between maximizing your performance metric and minimizing the number of steps in the pipeline. Thus, you should see a pipeline with only one classifier in it, but you should also see a better-performing pipeline with (for example) many preprocessing steps before a classifier.

**Comment by mattvan83** [comment]
@rhiever please find below the pareto_front_fitted_pipelines_ attribute of
the TPOT object with "TPOT_light" configuration:













*{'DecisionTreeClassifier(input_matrix,
DecisionTreeClassifier__criterion=gini,
DecisionTreeClassifier__max_depth=3,
DecisionTreeClassifier__min_samples_leaf=11,
DecisionTreeClassifier__min_samples_split=15)':
Pipeline(steps=[('decisiontreeclassifier',
DecisionTreeClassifier(class_weight=None, criterion='gini',
max_depth=3,            max_features=None, max_leaf_nodes=None,
min_impurity_split=1e-07, min_samples_leaf=11,
min_samples_split=15, min_weight_fraction_leaf=0.0,
presort=False, random_state=None, splitter='best'))]),
'LogisticRegression(DecisionTreeClassifier(input_matrix,
DecisionTreeClassifier__criterion=gini,
DecisionTreeClassifier__max_depth=10,
DecisionTreeClassifier__min_samples_leaf=18,
DecisionTreeClassifier__min_samples_split=3), LogisticRegression__C=10.0,
LogisticRegression__dual=DEFAULT, LogisticRegression__penalty=l1)':
Pipeline(steps=[('stackingestimator',
StackingEstimator(estimator=DecisionTreeClassifier(class_weight=None,
criterion='gini', max_depth=10,            max_features=None,
max_leaf_nodes=None,            min_impurity_split=1e-07,
min_samples_leaf=18,            min_samples_split=3,
min_weight_fraction_leaf=...ty='l1', random_state=None, solver='liblinear',
tol=0.0001,          verbose=0, warm_start=False))]),
'LogisticRegression(VarianceThreshold(DecisionTreeClassifier(input_matrix,
DecisionTreeClassifier__criterion=gini,
DecisionTreeClassifier__max_depth=10,
DecisionTreeClassifier__min_samples_leaf=18,
DecisionTreeClassifier__min_samples_split=3),
VarianceThreshold__threshold=0.7), LogisticRegression__C=10.0,
LogisticRegression__dual=DEFAULT, LogisticRegression__penalty=l1)':
Pipeline(steps=[('stackingestimator',
StackingEstimator(estimator=DecisionTreeClassifier(class_weight=None,
criterion='gini', max_depth=10,            max_features=None,
max_leaf_nodes=None,            min_impurity_split=1e-07,
min_samples_leaf=18,            min_samples_split=3,
min_weight_fraction_leaf=...ty='l1', random_state=None, solver='liblinear',
tol=0.0001,          verbose=0, warm_start=False))])}*

Could you help me understand in it where should I:
1) find the best-discovered pipelines that have a trade-off between
maximizing your performance metric and minimizing the number of steps in
the pipeline ?
2) find a better-performing pipeline with (for example) many preprocessing
steps before a classifier ?

Could you explain me the significance of for example
"*LogisticRegression(DecisionTreeClassifier(input_matrix,
DecisionTreeClassifier__criterion=gini,
DecisionTreeClassifier__max_depth=10,
DecisionTreeClassifier__min_samples_leaf=18,
DecisionTreeClassifier__min_samples_split=3), LogisticRegression__C=10.0,
LogisticRegression__dual=DEFAULT, LogisticRegression__penalty=l1)'" *where
*LogisticRegression* and *DecisionTreeClassifier* seems to be imbricated ?

Many thanks in advance !

Best regards,
Matthieu


2017-08-29 16:34 GMT+02:00 Randy Olson <notifications@github.com>:

**Comment by weixuanfu** [comment]
@mattvan83 

`pareto_front_fitted_pipelines_` attribute is a python dictionary containing the all pipelines on the TPOT Pareto front, where the key is the string representation of the pipeline and the value is the corresponding pipeline fitted on the entire training dataset. The TPOT Pareto front provides a trade-off between pipeline complexity (i.e., the number of steps in the pipeline) and the predictive performance of the pipeline. 

In the example in the comment above, the key is:
```
'LogisticRegression(DecisionTreeClassifier(input_matrix,
DecisionTreeClassifier__criterion=gini,
DecisionTreeClassifier__max_depth=10,
DecisionTreeClassifier__min_samples_leaf=18,
DecisionTreeClassifier__min_samples_split=3), LogisticRegression__C=10.0,
LogisticRegression__dual=DEFAULT, LogisticRegression__penalty=l1)'
``` 
and its value is:
```
Pipeline(steps=[('stackingestimator',
StackingEstimator(estimator=DecisionTreeClassifier(class_weight=None,
criterion='gini', max_depth=10,            max_features=None,
max_leaf_nodes=None,            min_impurity_split=1e-07,
min_samples_leaf=18,            min_samples_split=3,
min_weight_fraction_leaf=...ty='l1', random_state=None, solver='liblinear',
tol=0.0001,          verbose=0, warm_start=False))])
```
You may use the pipeline object directly for subsequent analysis as a [scikit-learn Pipeline object](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) .

However, the `print` function does not print all details in the pipeline if the pipeline is very complex. For better presentation of the pipeline, you could try the demo codes below:

```
from tpot import TPOTClassifier
from tpot.export_utils import expr_to_tree, generate_pipeline_code
from deap import creator
tpot_obj = TPOTClassifier()
# a demo key of your_TPOT_object.pareto_front_fitted_pipelines_
pipeline_string = (
      'LogisticRegression(DecisionTreeClassifier(input_matrix,'
      'DecisionTreeClassifier__criterion=gini,'
      'DecisionTreeClassifier__max_depth=10,'
      'DecisionTreeClassifier__min_samples_leaf=18,'
      'DecisionTreeClassifier__min_samples_split=3), LogisticRegression__C=10.0,'
      'LogisticRegression__dual=False, LogisticRegression__penalty=l1)'
)
# a deap object
optimized_pipeline = creator.Individual.from_string(pipeline_string, tpot_obj._pset)
# A string for scikit-learn Pipeline
sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(optimized_pipeline, tpot_obj._pset), tpot_obj.operators)
# scikit-learn Pipeline object
sklearn_pipeline_obj = eval(sklearn_pipeline_str, tpot_obj.operators_context)  

print(sklearn_pipeline_str)
``` 
And you also find the `StackingEstimator` transformer in the Pipeline object for stacking prediction from `DecisionTreeClassifier` in the first step of pipeline, and second step is `LogisticRegression`

**Comment by rhiever** [comment]
Closing this issue. Please feel free to re-open or file a new issue if you have any further questions or comments.
